Map layout

Nitwel opened the pull request:
I moved this feature into an internal branch for ease of reviewing and because Github doesn't really like rebasing / merging the current version of directus into the old PR. 😒
The initial PR can be found here: https://github.com/directus/directus/pull/3626

---

The library used is [maplibre-gl-js](https://github.com/maplibre/maplibre-gl-js), the community fork of [mapbox-gl-js](https://github.com/mapbox/mapbox-gl-js). I picked the former due to the new restrictive license of mapbox-gl v2. Mapbox is not the most feature rich library, but it offers the best performance for displaying huge dataset thanks to it being based on webGL, and having a built-in clustering and geometry simplification. Its is the library used in Apache Superset.

Converting raw output to GeoJSON is done in web worker with the help of [comlink](https://github.com/GoogleChromeLabs/comlink) and [worker-plugin](https://github.com/GoogleChromeLabs/worker-plugin).

Screenshots:

![Screenshot 2021-01-09 011114](https://user-images.githubusercontent.com/33065839/104078271-7e371380-521d-11eb-861d-08b9c0ba5027.png)

Features:

- [x] Supported format: (e|t)WKB ( Postgis ), (e)WKT, GeoJSON, Longitude / Latitude

- [x] Support for any feature type (Point, Polygon, Line, ...)

- [x] Basemaps from Carto and Esri

- [x] Automatic view fit to data bounds

- [x] Link to edit page on feature click

- [x] GeoJSON serializing progress bar

- [x] Clustering options

- [x] Animation options

- [x] Style customisation

- [x] Select mode


- [x] Support for other SRID (currently only WGS84)

- [x] Data preview on hover (help needed)

Suggestions:

- Since no scroll is possible, it would be nice to be able to set header bar in its collapsed state programatically, by making that state a store for example (like the sidebar). Also the header bar is currently hard coded to 64px. In order to ensure that the map layout will always cover 100% of the screen even if that number were to change, it would be better if that value were a css variable.

Issues:

- Sometimes on collection change, the server crash with `INTERNAL_SERVER_ERROR`. The code is returned after the first `GET` request for the new collection. It seems to be caused by the fact that the url parameter for `sort` is still set to the previous collection one. The crash would then happen because you couldn't sort a collection over a field it doesn't have. I don't know if it is a bug with directus or with my code.

- Looking for a better looking pin icon with shadow.

Looking for any insights on this, suggestions are welcome! Cheers

oreilles committed:
Added map layout

oreilles committed:
Cleanup and bug fixes

oreilles committed:
Removed package-lock

oreilles committed:
Cleanup and fixes

oreilles committed:
Small fix

oreilles committed:
Added back package-lock

oreilles committed:
Saved camera, autofitting option, bug fixes

oreilles committed:
Refactor and ui improvements

oreilles committed:
Improvements

oreilles committed:
Added seled mode

oreilles committed:
Removed unused dependency

oreilles committed:
Changed selection behaviour, cleanup.

Nitwel committed:
update import and dependencies

benhaynes commented:
What does the current pin/marker icon look like? If you give me the specs I can create one! SVG/PNG? Size? Where is the registration point?

Nitwel commented:
Just looks like a dot.
![grafik](https://user-images.githubusercontent.com/22577866/113308155-30a2e200-9306-11eb-8cc5-0770633c5eff.png)

benhaynes commented:
Ahh, gotcha. I actually don't mind that... can we update the color programmatically though?

Nitwel committed:
make custom style into drawer

Nitwel commented on app/src/layouts/map/worker/index.ts:
```diff
@@ -0,0 +1,118 @@
+	options: geometryOptions,
+	onProgress: (p: number) => void
+): GeoJSON.FeatureCollection {
+	const parser = getGeometryParser(options);
+	const project = (coord: any): any => coord; // TODO
+	const geojson: GeoJSON.FeatureCollection = {
```
@Oreilles What where you planning to do here?

Oreilles replied:
Handling the te projection of coordinates from any CRS to WGS84, which is the standard for GeoJSON, using the library proj4js.

Nitwel replied:
And why? 😅

Oreilles replied:
Because not everyone use WGS84 😉 There are many countries in which public geospatial data is distributed in the CRS that makes the most sense for said country. In France for example, most datasets provided by the IGN are in EPSG:2154 because it allows for the least amount of distortion over France when viewed on a 2D map.

Nitwel replied:
Is this a feature that you think is nessesary before we're going to merge this?
Also, is there a way to find out what projection was used or does this has to be configured by the user?

Oreilles replied:
There isn't a way to guess from the data only (there are thousands of different CRS), so it should indeed be configured by the user (defaulting to WGS84 if unspecified). I think it's a very important feature, as geospatial data is basically unusable if in a different CRS (it will be shown on the map at wrong locations, or nowhere at all if the projections are too different) - plus it's not that much work to implement as there already exist a library doing that (proj4). 

But it's true that most non-specialists only deal with WGS84 data, so ultimately I don't think it is absolutely necessary for a v1.

Nitwel committed:
remove unused imports

Oreilles commented:
> Ahh, gotcha. I actually don't mind that... can we update the color programmatically though?

You can change the style of every feature (not just [points](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#circle)) within a JSON field in the option sidebar, according to the [mapbox style spec](https://docs.mapbox.com/mapbox-gl-js/style-spec/).

The thing about that spec is that if you want to use icons, you have to supply an JSON index file, containing locations and sizes of icons in a PNG spritesheet. Those two files must be hosted as you can only provide a url link to them. As far as I'm aware, only [openmaptiles](https://github.com/openmaptiles/osm-bright-gl-style/blob/master/style.json) and [esri](https://www.arcgis.com/sharing/rest/content/items/f455d84d5aab4d64b1191e0d2eed42b1/resources/styles/root.json?f=pjson) provides a free to use complete set of icons that way, but I wasn't impressed by both of their marker icon, so I didn't bother using it. There's actually [another way](https://docs.mapbox.com/mapbox-gl-js/example/add-image/) to use icons, but it's kind of clunky, and would be pretty hard for end user to customize.

Writing a GUI to edit the style file is a massive task as the spec is really long ([maputnik](https://maputnik.github.io/editor/#0.83/0/0) did it, I believe this is the tool people should use in order to create the style they want).

As per this branch, I have some changes to push soon! Some refactoring, changes to the select mode, and a props preview on hover.

Nitwel commented:
> As per this branch, I have some changes to push soon! Some refactoring, changes to the select mode, and a props preview on hover.

Awesome to hear! 😄 
You could either push these changes to your old branch of your fork and then I merge them onto this branch or we would have to ask @rijkvanzanten so he gives you access to this branch.

Oreilles commented:
@benhaynes Since this is about to come into the main branch soon, where is the conversation we had about "full page layout" at ? 🙂

benhaynes commented:
Since @Nitwel is taking on this PR, and already mentioned the header needing to _start_ in the "smaller" state, hopefully he can also push a header prop that allows this to force that. Nils, any issues with that?

Nitwel committed:
use lodash functions

Oreilles commented on app/src/layouts/map/maps.vue:
```diff
@@ -233,31 +233,7 @@ import { useRelationsStore } from '../../stores/';
-
-function assign(a: any, b: any): any {
-	if (![a, b].every(isObject)) return (a = b);
-	if ([a, b].every(isArray)) return a.concat(b);
```
This line is why I didn't use `lodash.merge`. When mergeing two arrays, they need to be concatenated but instead lodash merge their values by index.

Nitwel replied:
When do we have the case that `userLayers.value` is an array instead of an object? I would like to keep everything as simple as possible.

Oreilles replied:
Oh, nevermind. It was only needed in a previous version where users could add custom sources through the json field, but not anymore since I removed it. Sorry about that.

Nitwel committed:
add popups

Oreilles commented on app/src/layouts/map/maps.vue:
```diff
@@ -457,7 +433,7 @@ export default defineComponent({
 					geojsonProgress.value = 0;
 					geojsonError.value = null;
-					geojson.value = await workerProxy(items.value, _layoutOptions.value, proxy(onProgress));
+					geojson.value = await workerProxy(items.value, _layoutOptions.value, '', proxy(onProgress));
```
Not sure about this..!

Nitwel replied:
Haha, I was just testing how to make custom templates work. 😄

Nitwel committed:
allow header to become small

Nitwel committed:
reorganize settings

Nitwel committed:
add styling to popup

Nitwel committed:
change default template

Nitwel commented:
Really enjoing reviewing this. And the fact that it can handle so many items is just crazy!
Awesome work @Oreilles ! ❤️ 

> As per this branch, I have some changes to push soon! Some refactoring, changes to the select mode, and a props preview on hover.

Also sorry if you also already did implement popups, really want to get this done so I added them myself.

![grafik](https://user-images.githubusercontent.com/22577866/113416563-e54e0980-93c1-11eb-9231-83fea2b21bb9.png)

Nitwel committed:
add projection option

Oreilles commented:
This layout is using maplibre-gl, but the map interface proposed in https://github.com/directus/directus/pull/3183 is using leaflet. Also, the map interface only allows selecting unique points, wherehas the layout can display any kind of geometry. For the sake of bundle size and consistency, I think it would be better if both components were using the same map library and the same background maps.

There's an extension maintained by mapbox itself, [mapbox-gl-draw]( https://github.com/mapbox/mapbox-gl-draw), that allows creating and editing (multi) points, lines, and polygons, which could be used for that purpose.

Nitwel commented:
I agree that it would be nice to use the same library in both interface and layout but that isn't our goal right now. Our goal is to be able to manage locations at all and nothing more.

rijkvanzanten commented:
> I agree that it would be nice to use the same library in both interface and layout but that isn't our goal right now. Our goal is to be able to manage locations at all and nothing more.

I disagree @Nitwel. No need to incur any technical debt in there now. @Oreilles is right. We should make sure that the layout and interface rely on the same underlying technologies for improved consistency, performance, and compatibility.

Nitwel committed:
add basic map interface

Nitwel committed:
finish simple map

Nitwel committed:
add mapbox style

Nitwel committed:
support more mapbox layouts

Nitwel committed:
add api key option

Oreilles commented:
Do we really want to add Mapbox styles ? Esri provides free ones based on OSM data that don't require an API key.

Nitwel committed:
add mapbox backgrounds to layout

Nitwel committed:
warn when no api key is set

Nitwel commented:
Why not? They're optional so for those who want to use them can get an key for and everybody else uses the free ones.
And they're in SVG format wich is really nice!

Nitwel committed:
Merge branch 'main' into map-layout

Oreilles commented:
> Why not? They're optional so for those who want to use them can get an key for and everybody else uses the free ones.
> And they're in SVG format wich is really nice!

It's great that you added the support for them! But IMO as they're not free, they shouldn't be part of our pre-selection. Just like we can customize layers, maybe we could just add the option to customize the sources available in the dropdown ?

Nitwel committed:
fix for latest version

jroxendal commented:
Here's the usage policy for the OSM tile server:
https://operations.osmfoundation.org/policies/tiles/

Doesn't seem like something you'd wanna hard code into Directus, though I'm sure not too many people are actually using the geospatial features as of yet. 

I'm sure there are tons of services like https://cloud.maptiler.com/maps/ where you can get an API key with a free account. Also, people publish their keys rather carelessly around the web if you wanna _borrow_ one:
https://codepen.io/search/pens?q=cloud.maptiler.com ;)

Being forced to get a free account shouldn't be too big of a hurdle for most people. If you choose to release it with the OSM tile layer, at least an admin should be nagged into switching provider they use it too much. 

Another good mitigation would be to implement a caching proxy in Directus, so that at least the default view is cached. If your users don't pan/zoom around too much, most tiles will be cached an the tile server won't be hit.

zenati commented:
@rijkvanzanten Is it a good idea to have mapbox_key in the settings table ? Usually api_keys are in ENV instead of DB.

https://github.com/directus/directus/pull/4805/files#diff-4762df17ac2feee8e681b2a7e1cc811ade834959215b2d766813d24ee742e615R5

What do you think ? What will happen if GoogleMaps API_KEY is added in future ?

oreilles committed:
Improved map layout and interface, bug fixes, refactoring.

.

.

oreilles committed:
Added postgis geometry format, added marker icon shadow

rijkvanzanten commented:
Closed in favor of https://github.com/directus/directus/pull/5038#issuecomment-819368008

oreilles committed:
Made map buttons bigger and their icons thinner. Added transition to header bar.

oreilles committed:
Bug fixes and error handling in map interface.

oreilles committed:
Moved box-select control out of the map component. Removed material icons sprite and use addImage for marker support.

oreilles committed:
Handle MultiGeometry -> Geometry interface error.

oreilles committed:
Removed hardcoded styles. Added migrations for basemap column. Lots of refactoring.

Removed hardcoded styles. Added migrations for basemap column. Lots of refactoring.

oreilles committed:
Fixed style reloading error. Added translations.

oreilles committed:
Moved worker code to lib.

oreilles committed:
Removed worker code. Prevent Mapbox from removing access_token from the URL.

oreilles committed:
Refactoring.

oreilles committed:
Change basemap selection to in-map dropdown for layout and interface.

oreilles committed:
Touchscreen selection support and small fixes.

oreilles committed:
Small change.

danieljwestman commented:
Whats the current status of this fantastic layout? An humble question ☺️

Oreilles commented:
@danieljwestman Discussion is ongoing in https://github.com/directus/directus/pull/5038, but the layout is very much functional. If you want to try it and give us your insights, just clone or fork it from [here](https://github.com/Oreilles/directus/tree/map-layout) 🙂

oreilles committed:
Merge rc.62 into map-layout-next

oreilles committed:
Fixed unused imports.

oreilles committed:
Added support for PostgreSQL identity column

oreilles committed:
Merge branch 'postgres-identity-column' into map-layout-next

oreilles committed:
Renamed migration. Added crs translation.

oreilles committed:
Only show fields using the map interface in the map layout.

oreilles committed:
Merge branch 'rc.64' into map-layout-next

# Conflicts:
#	app/package.json
#	app/src/lang/translations/en-US.yaml
#	package-lock.json

oreilles committed:
Only show fields using the map interface in the map layout. Bumped to v9.0.0-rc.64.

oreilles committed:
Removed logging.

oreilles committed:
Reverted Dockerfile change.

oreilles committed:
Improved crs support.

oreilles committed:
Fixed translations.

oreilles committed:
Check for schema identity before updating it.

oreilles committed:
Fixed popup not updating on feature hover.

oreilles committed:
Merge branch 'prevent-unnecessary-schema-update' into map-layout

oreilles committed:
Added feature hover styling. Fixed layer customization input. Added out of bounds error handling.

oreilles committed:
Added geometry type and support for database native geometries.

rijkvanzanten committed:
Merge branch 'main' into map-layout

oreilles committed:
Merge remote-tracking branch 'upstream/main' into map-layout

oreilles committed:
Fixed linting.

oreilles committed:
Merge branch 'map-layout' of github.com:Oreilles/directus into map-layout

oreilles committed:
Fixed layout.

oreilles committed:
Fixed layout.

oreilles committed:
Merge branch 'main' into map-layout

oreilles committed:
Actually fixed linting

oreilles committed:
Merge remote-tracking branch 'upstream/main' into map-layout

oreilles committed:
Merge remote-tracking branch 'upstream/main' into map-layout

oreilles committed:
Full support for native geometries
Fixed basemap input
Improved feature popup on hover
Locked interfaced support

oreilles committed:
Merged upstream

oreilles committed:
Fixed geometryType option not updating

oreilles committed:
Bug fixes in interface

oreilles committed:
Merge remote-tracking branch 'upstream/main' into map-layout

oreilles committed:
Fixed crash when empty basemap settings. Fixed fitBounds option not updating.

oreilles committed:
Added back storage type option. Improved interface behaviour.

oreilles committed:
Dropped wkb because of vendor inconsistency with binary data

oreilles committed:
Updated layout to match new geometry type. Fixed geojson payload transform.

oreilles committed:
Added missing geometry_format attributes to local types.

oreilles committed:
Fixed typos & refactoring

oreilles committed:
Removed dependency on proj4

oreilles committed:
Fix error when empty map interface options

oreilles committed:
Set geometry SRID to 4326 when inserting into the database

oreilles committed:
Add support for selectMode

oreilles committed:
Fix error on initial source load

oreilles committed:
Merge remote-tracking branch 'upstream/main' into map-layout

oreilles committed:
Merge remote-tracking branch 'upstream/main' into map-layout

rijkvanzanten committed:
Use postgis instead of postgres for DB image
